File: Tyche.iOS/UI/Sources/UI/EditableViewState.swift
public enum EditableViewState<Value> {
    case initial(Value)
    case loading(current: Value, target: Value)
    case success(current: Value, succeeded: Value)
    case failure(current: Value, failed: Value, error: Error)
}

public extension EditableViewState {
    @inlinable
    func isInitial() -> Bool {
        if case .initial = self {
            return true
        }
        return false
    }
    
    @inlinable
    func isLoading() -> Bool {
        if case .loading = self {
            return true
        }
        return false
    }
    
    @inlinable
    func isSuccess() -> Bool {
        if case .success = self {
            return true
        }
        return false
    }
    
    @inlinable
    func isFailure() -> Bool {
        if case .failure = self {
            return true
        }
        return false
    }
    
    @inlinable
    func value() -> Value {
        return switch self {
        case .initial(let value), .loading(current: let value, _), .success(_, succeeded: let value), .failure(current: let value, _, _):
            value
        }
    }
    
    @inlinable
    func valueOrNull() -> Value? {
        return switch self {
        case .initial(let value), .loading(current: let value, _), .success(_, succeeded: let value):
            value
        default:
            nil
        }
    }
    
    @inlinable
    func errorOrNull() -> Error? {
        if case .failure(_, _, let error) = self {
            return error
        }
        return nil
    }
}

File: Tyche.Android/ui/src/main/java/com/felipearpa/tyche/ui/state/EditableViewState.kt
package com.felipearpa.tyche.ui.state

sealed class EditableViewState<out Value : Any> {
    data class Initial<Value : Any>(val value: Value) : EditableViewState<Value>()

    data class Loading<Value : Any>(val current: Value, val target: Value) :
        EditableViewState<Value>()

    data class Success<Value : Any>(val old: Value, val succeeded: Value) :
        EditableViewState<Value>()

    data class Failure<Value : Any>(
        val current: Value,
        val failed: Value,
        val exception: Throwable
    ) : EditableViewState<Value>()
}

fun <Value : Any> EditableViewState<Value>.isInitial() = this is EditableViewState.Initial

fun <Value : Any> EditableViewState<Value>.isLoading() = this is EditableViewState.Loading

fun <Value : Any> EditableViewState<Value>.isSuccess() = this is EditableViewState.Success

fun <Value : Any> EditableViewState<Value>.isFailure() = this is EditableViewState.Failure

fun <Value : Any> EditableViewState<Value>.exceptionOrNull(): Throwable? {
    return when (this) {
        is EditableViewState.Failure -> this.exception
        else -> null
    }
}

fun <Value : Any> EditableViewState<Value>.currentValue() =
    when (this) {
        is EditableViewState.Initial -> this.value
        is EditableViewState.Loading -> this.current
        is EditableViewState.Success -> this.succeeded
        is EditableViewState.Failure -> this.current
    }